cql3/multi_column_relation.hh

```diff
@@ -145,7 +145,10 @@ protected:
             return cs->column_specification;
         });
         auto t = to_term(col_specs, *get_value(), db, schema->ks_name(), bound_names);
```

Comment:
> This begins the addition of a new restrictions representation, which
will continue over the next several patches.  As the new
representation is built, it will be compared for equivalence with the
old one.  When the results from both are identical, the old one will
be removed.

>The new representation consists of several new classes in the
cql3::restrictions::wip namespace.  They will be constructed in
parallel with existing restrictions and stored together with them.  In
time, their functionality will grow to match the old representation.
When the old representation is finally removed, these new classes will
move out of the wip namespace.

>To replicate the existing functionality, we will gradually add free
functions that implement restriction methods such as bounds,
has_function, or is_satisfied_by.  As each such function is added, it
will be invoked together with its original equivalent and the results
of both compared for accuracy.

Consider the approach taken by @espindola in 20aed3398de28a38c79287145af6d79a9861b6eb. Instead of constructing a parallel hierarchy, he morphs the existing heirarchy one function at a time by moving it into a variant invocation. This allows verification by inspection rather than execution (which may not cover all paths) and doesn't cause performance degradations, allowing it to be merged.

Reply:
From a quick glance at 20aed33, it morphs the hierarchy _members_, not the hierarchy itself.  This change, OTOH, aims to completely re-engineer a hierarchy infamous for its unreadability, weird limitations, and a messy extension procedure.

The parallel-hierarchy approach was suggested by @psarna in a design document I'll share with you shortly.  He justified it by noting that the current restrictions code is a result of long and painful evolution to fit incredibly subtle requirements.  Any change to the current code risks introducing bugs that history shows are difficult to find and fix.  (Not to mention that the current code makes incremental changes onerously expensive -- the very reason we want to rewrite it.)  Better to construct the new hierarchy in parallel and compare its intermediate results to those of the old one, catching any discrepancies early on.

Reply:
Given the regressions already discovered it doesn't look successful. We are throwing away the existing knowledge embedded and the code, and hoping to discover problems through testing.

Reply:
> We are throwing away the existing knowledge embedded and the code,

We are not, though -- parallel structure means precisely that we're keeping them around for a while.  Existing code will be replicated as much as possible in new code, and tests (old and new) will ensure their equivalence before anything is thrown out.

But we cannot leave the class hierarchy and restriction-creation procedures as they currently are.  They have terrible maintainability, extensibility, and limitations (eg, #4344) that preclude matching C* and other interfaces we aim to implement.  Merely replacing their methods with free functions will not solve these problems.  It won't even save labour in the long run.

My plan is to replace the `restrictions` class with `expression` as the first step, then go on replacing other numerous restriction classes one by one.  We don't need to merge this until actual replacement occurs and the old class is dropped.

Reply:
Oh, one more explanation:
> Given the regressions already discovered it doesn't look successful

The new code has not yet been turned on in existing tests.  That's an integral part of @psarna's proposal and will happen once all `restriction` subtypes are implemented in wip.  Until I do that, condemning the proposal is premature.

Reply:
The knowledge is lost, and you hope that we'll regain it through parallel testing. No doubt we'll gain some of it back from tests failing, and no doubt review will find bugs, but there is a source of knowledge that we're not using and it may end up causing regressions.

Reply:
This is true, and I acknowledge the risk.  But how can we avoid losing old knowledge if we intend to clean up the problematic code?

Reply:
The traditional method (very hard in such cases) is to convert rather than do side-by-side.

A lesser method is to locate the existing code that is responsible for the same thing, and write the new code using it as a guide. This is less reviewer friendly as it doesn't leave a trace. Maybe you can add comments pointing to the original code.